### PR TITLE
Ledgers without masterkey in metadata can't be recovered

### PR DESCRIPTION
**Pull request by ivankelly**
Test that ledgers, created without a master key in the metadata, cannot be recovered until
the master key is known.

Replaces TestBookieRecovery from the old tests.

Master Issue: #903

### CODE REVIEW

File: tests/backward-compat-recovery-no-password/src/test/groovy/org/apache/bookkeeper/tests/TestCompatRecoveryNoPassword.groovy [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+            bka.close();
+        }
+
+        adminConf.setBookieRecoveryDigestType(BookKeeper.DigestType.CRC32)
+        adminConf.setBookieRecoveryPasswd(passwdCorrect)
+
+        bka = new BookKeeperAdmin(adminConf)
+        try {
+            bka.recoverBookieData(failedBookieId)
+            Assert.fail("Shouldn't be able to recover with wrong digest")
```
Comment (eolivelli) [inline tests/backward-compat-recovery-no-password/src/test/groovy/org/apache/bookkeeper/tests/TestCompatRecoveryNoPassword.groovy]: I don't know if we can use 'import static' in Groovy.

I wonder if we can have some flavour of checkstyle for Groovy

/cc @sijie

File: tests/backward-compat-recovery-no-password/src/test/groovy/org/apache/bookkeeper/tests/TestCompatRecoveryNoPassword.groovy [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+        byte[] passwdCorrect = "AAAAAA".getBytes()
+        byte[] passwdBad = "BBBBBB".getBytes()
+
+        String currentVersion = System.getProperty("currentVersion")
+        String zookeeper = BookKeeperClusterUtils.zookeeperConnectString(docker)
+
+        BookKeeperClusterUtils.legacyMetadataFormat(docker)
+
+        // Create a 4.1.0 client, will update /ledgers/LAYOUT
+        def v410CL = MavenClassLoader.forBookKeeperVersion("4.1.0")
```
Comment (eolivelli) [inline tests/backward-compat-recovery-no-password/src/test/groovy/org/apache/bookkeeper/tests/TestCompatRecoveryNoPassword.groovy]: I think we should try to close all classloaders

If I undersand correctly we are using forkCount = 0, so no new JVM will be spawned for test.
We should take care of not leaving memory (classloader) leaks

File: tests/backward-compat-recovery-no-password/src/test/groovy/org/apache/bookkeeper/tests/TestCompatRecoveryNoPassword.groovy (hunk 1 above)
Comment (sijie) [reply tests/backward-compat-recovery-no-password/src/test/groovy/org/apache/bookkeeper/tests/TestCompatRecoveryNoPassword.groovy]: I don't think we enforce any checkstyle about import static. We only disallow "import *". We are okay to use "Assert.fail" or "import static ..fail;" and use "fail". 

Also I don't know how checkstyle is applied to groovy. I think we need an issue to enforce checkstyle on groovy as well.

File: tests/backward-compat-recovery-no-password/src/test/groovy/org/apache/bookkeeper/tests/TestCompatRecoveryNoPassword.groovy (hunk 2 above)
Comment (ivankelly) [reply tests/backward-compat-recovery-no-password/src/test/groovy/org/apache/bookkeeper/tests/TestCompatRecoveryNoPassword.groovy]: Added a note to #1030

File: tests/backward-compat-recovery-no-password/src/test/groovy/org/apache/bookkeeper/tests/TestCompatRecoveryNoPassword.groovy (hunk 1 above)
Comment (ivankelly) [reply tests/backward-compat-recovery-no-password/src/test/groovy/org/apache/bookkeeper/tests/TestCompatRecoveryNoPassword.groovy]: Create #1043

**Comment by ivankelly** [comment]
@eolivelli do you have more comments, or can I merge?

**Pull request closed without merging**